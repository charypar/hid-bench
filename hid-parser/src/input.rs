use std::fmt::Display;

// Represents a single input item in a report
#[derive(Debug)]
pub struct Input {
    pub usage: (u16, u16),
    pub value: InputValue,
}

#[derive(Debug)]
pub enum InputValue {
    Bool(bool),
    UInt(u32),
    Int(i32),
    None, // "Null state"
}

impl Display for Input {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.value {
            InputValue::Bool(b) => write!(f, "({:02x} {:02x}): {}", self.usage.0, self.usage.1, b),
            InputValue::UInt(u) => write!(f, "({:02x} {:02x}): {}", self.usage.0, self.usage.1, u),
            InputValue::Int(i) => write!(f, "({:02x} {:02x}): {}", self.usage.0, self.usage.1, i),
            InputValue::None => write!(f, "None"),
        }
    }
}
