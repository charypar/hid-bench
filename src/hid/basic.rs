// 1st level: Parse basic items

#[derive(Debug)]
pub struct BasicItems<'a> {
    bytes: &'a [u8],
    offset: usize,
}

impl<'a> BasicItems<'a> {
    pub fn new(bytes: &'a [u8]) -> Self {
        BasicItems { bytes, offset: 0 }
    }
}

impl<'a> BasicItems<'a> {
    fn item_header(header: u8) -> (usize, u8, u8) {
        let sizes: [usize; 4] = [0, 1, 2, 4];

        let size = sizes[(header & 0b11) as usize];
        let item_type = (header & 0b1100) >> 2;
        let tag = (header & 0b11110000) >> 4;

        (size, item_type, tag)
    }
}

impl<'a> Iterator for BasicItems<'a> {
    type Item = BasicItem;

    fn next(&mut self) -> Option<Self::Item> {
        if self.offset >= self.bytes.len() {
            return None;
        }

        let (size, item_type, tag) = Self::item_header(self.bytes[self.offset]);

        let mut data = 0u32;
        for byte_idx in 0..size {
            // build up from little-endian ordered bytes
            data |= (self.bytes[self.offset + 1 + byte_idx] as u32) << (byte_idx * 8);
        }

        self.offset += size + 1;

        Some(BasicItem::new(item_type, tag, data, size))
    }
}

// NOTE only short items are supported
#[derive(Debug)]
pub enum BasicItem {
    Main(MainItem),
    Global(GlobalItem),
    Local(LocalItem),
    Reserved,
}

impl BasicItem {
    fn new(item_type: u8, tag: u8, data: u32, size: usize) -> Self {
        match item_type {
            0 => Self::Main(MainItem::new(tag, data)),
            1 => Self::Global(GlobalItem::new(tag, data)),
            2 => Self::Local(LocalItem::new(tag, data, size)),
            _ => Self::Reserved,
        }
    }
}

#[derive(Debug)]
pub enum MainItem {
    Input(InputItemData),
    Output(OutputItemData),
    Feature(FeatureItemData),
    Collection(Collection), // TODO work out collection type representation
    EndCollection,
    Reserved,
}

impl MainItem {
    fn new(tag: u8, data: u32) -> Self {
        match tag {
            0b1000 => Self::Input(InputItemData { data }),
            0b1001 => Self::Output(OutputItemData { data }),
            0b1011 => Self::Feature(FeatureItemData { data }),
            0b1010 => Self::Collection(Collection::new(data as u8)),
            0b1100 => Self::EndCollection,
            _ => Self::Reserved,
        }
    }
}

#[derive(Debug)]
pub struct InputItemData {
    pub data: u32,
}

#[derive(Debug)]
pub struct OutputItemData {
    pub data: u32,
}

#[derive(Debug)]
pub struct FeatureItemData {
    pub data: u32,
}

#[derive(Debug, Clone, Copy)]
pub enum Collection {
    Physical,
    Application,
    Logical,
    Report,
    NamedArray,
    UsageSwitch,
    UsageModifier,
    Reserved,
    Vendor(u8),
}

impl Collection {
    fn new(data: u8) -> Self {
        match data {
            0 => Self::Physical,
            1 => Self::Application,
            2 => Self::Logical,
            3 => Self::Report,
            4 => Self::NamedArray,
            5 => Self::UsageSwitch,
            6 => Self::UsageModifier,
            n if (7..0x7F).contains(&n) => Self::Reserved,
            n => Self::Vendor(n),
        }
    }
}

#[derive(Debug)]
pub enum GlobalItem {
    UsagePage(u16),
    LogicalMinimum(i32),
    LogicalMaximum(i32),
    PhysicalMinimum(i32),
    PhysicalMaximum(i32),
    UnitExponent(u32),
    Unit(u32), // TODO decode
    ReportSize(u32),
    ReportID(u8),
    ReportCount(u32),
    Push,
    Pop,
    Reserved,
}

impl GlobalItem {
    fn new(tag: u8, data: u32) -> Self {
        match tag {
            0 => Self::UsagePage(data as u16),
            1 => Self::LogicalMinimum(data as i32), // FIXME check this works with signs
            2 => Self::LogicalMaximum(data as i32), // FIXME check this works with signs
            3 => Self::PhysicalMinimum(data as i32), // FIXME check this works with signs
            4 => Self::PhysicalMaximum(data as i32), // FIXME check this works with signs
            5 => Self::UnitExponent(data as u32),
            6 => Self::Unit(data),
            7 => Self::ReportSize(data),
            8 => Self::ReportID(data as u8),
            9 => Self::ReportCount(data),
            10 => Self::Push,
            11 => Self::Pop,
            _ => Self::Reserved,
        }
    }
}

#[derive(Debug)]
pub enum LocalItem {
    Usage(u16),
    UsageMinimum(u16),
    UsageMaximum(u16),
    ExtendedUsage(u16, u16),
    ExtendedUsageMinimum(u16, u16),
    ExtendedUsageMaximum(u16, u16),
    DesignatorIndex(u32),
    DesignatorMinimum(u32),
    DesignatorMaximum(u32),
    StringIndex(u32),
    StringMinimum(u32),
    StringMaximum(u32),
    Delimiter(bool), // true - open, false - close
    Reserved,
}

impl LocalItem {
    fn new(tag: u8, data: u32, size: usize) -> Self {
        match (tag, size) {
            (0, 1) => Self::Usage(data as u16),
            (0, 2) => Self::Usage(data as u16),
            (0, 4) => Self::ExtendedUsage((data >> 16) as u16, (data & 0xFF) as u16),
            (1, 1) => Self::UsageMinimum(data as u16),
            (1, 2) => Self::UsageMinimum(data as u16),
            (1, 4) => Self::ExtendedUsageMinimum((data >> 16) as u16, (data & 0xFF) as u16),
            (2, 1) => Self::UsageMaximum(data as u16),
            (2, 2) => Self::UsageMaximum(data as u16),
            (2, 4) => Self::ExtendedUsageMaximum((data >> 16) as u16, (data & 0xFF) as u16),
            (3, _) => Self::DesignatorIndex(data),
            (4, _) => Self::DesignatorMinimum(data),
            (5, _) => Self::DesignatorMaximum(data),
            (6, _) => Self::StringIndex(data),
            (7, _) => Self::StringMinimum(data),
            (8, _) => Self::StringMaximum(data),
            (9, _) => Self::Delimiter(data != 0),
            (_, _) => Self::Reserved,
        }
    }
}
