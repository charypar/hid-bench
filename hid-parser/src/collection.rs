use std::fmt::Display;

use super::input::Input;

// Collection type, reused for reports
#[derive(Debug)]
pub struct Collection<T> {
    pub collection_type: super::basic::Collection,
    pub usage: (u16, u16),
    // "String and Physical indices, as well as delimiters may be associated with collections."
    // TODO delimiter support (when needed)
    pub designator_index: Option<u32>,
    pub string_index: Option<u32>,
    pub items: Vec<CollectionItem<T>>,
}

impl<T> Collection<T> {
    pub fn map<O, F: Fn(&T) -> Option<O>>(&self, f: F) -> Collection<O>
    where
        F: Copy,
    {
        Collection {
            collection_type: self.collection_type,
            usage: self.usage,
            designator_index: self.designator_index,
            string_index: self.string_index,
            items: self
                .items
                .iter()
                .filter_map(|item| match item {
                    CollectionItem::Collection(c) => {
                        let col = c.map(f);

                        Some(CollectionItem::Collection(col))
                    }
                    CollectionItem::Item(item) => f(item).map(CollectionItem::Item),
                })
                .collect(),
        }
    }
}

#[derive(Debug)]
pub enum CollectionItem<T> {
    Collection(Collection<T>),
    Item(T),
}

impl Display for Collection<Vec<Input>> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let items_string = self
            .items
            .iter()
            .map(|item| match item {
                CollectionItem::Collection(c) => format!("{}", c),
                CollectionItem::Item(inputs) => format!(
                    "[{}]",
                    inputs
                        .iter()
                        .map(|input| format!("{}", input))
                        .collect::<Vec<_>>()
                        .join(", ")
                ),
            })
            .collect::<Vec<_>>()
            .join(", ");

        write!(
            f,
            "{:?}({:02x} {:02x})[{}]",
            self.collection_type, self.usage.0, self.usage.1, items_string
        )
    }
}
