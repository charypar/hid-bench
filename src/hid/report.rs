use super::input::{Input, InputValue};

// A single report, may read multiple inputs of the same configuration
#[derive(Debug)]
pub struct Report {
    pub report_type: ReportType,
    pub usages: Vec<(u16, u16)>,
    pub usage_minimum: Option<(u16, u16)>,
    pub usage_maximum: Option<(u16, u16)>,
    pub logical_minimum: i32,
    pub logical_maximum: i32,
    pub physical_minimum: i32,
    pub physical_maximum: i32,
    pub unit: Option<u32>,
    pub unit_exponent: Option<u32>,
    pub bit_offset: usize,     // start of the report in the overall report data
    pub report_id: Option<u8>, // if given, add 8 bits to the offset, check the ID matches
    pub report_size: u32,
    pub report_count: u32,
}

impl Report {
    pub fn parse(&self, report: &[u8]) -> Vec<Input> {
        let ReportType::Input(flags) = self.report_type;
        if (flags & 1) == 1 {
            return vec![];
        }

        let spec_usages = self.usages.len();

        (0..(self.report_count as usize))
            .map(|i| {
                let usage = if i < spec_usages {
                    self.usages[i]
                } else {
                    // Usage Minimum specifies the usage to be associated with the first unassociated control
                    // in the array or bitmap. Usage Maximum specifies the end of the range of usage values
                    // to be associated with item elements.
                    if let Some((up, u)) = self.usage_minimum {
                        (up, u + spec_usages as u16 + i as u16)
                    } else {
                        // HID 1.11, section 6.2.2.8 Local Items
                        //
                        // While Local items do not carry over to the next Main item,
                        // they may apply to more than one control within a single item.
                        // For example, if an Input item defining five controls is
                        // preceded by three Usage tags, the three usages would be
                        // assigned sequentially to the first three controls, and the
                        // third usage would also be assigned to the fourth and fifth controls.
                        self.usages[self.usages.len() - 1]
                    }
                };

                let offset = self.bit_offset + (self.report_size as usize * i);
                let base_value = Self::extract_value(report, offset, self.report_size);

                let value = match (self.logical_minimum, self.logical_maximum) {
                    (0, 1) => InputValue::Bool(base_value != 0),
                    (a, b) if a >= 0 && b >= 0 => InputValue::UInt(base_value),
                    _ => InputValue::Int(Self::signed(base_value, self.report_size)),
                };

                Input { usage, value }
            })
            .collect()
    }

    fn signed(value: u32, length: u32) -> i32 {
        let sign_mask = 1 << (length - 1);
        let number_mask = !(0xFFFF_FFFF << (length - 1));

        let sign = value & sign_mask;
        let unsinged_number = value & number_mask;

        if sign != 0 {
            (unsinged_number | !number_mask) as i32
        } else {
            unsinged_number as i32
        }
    }

    fn extract_value(report: &[u8], bit_offset: usize, bit_length: u32) -> u32 {
        let first_byte = bit_offset / 8; // first byte in which the value is
        let last_byte = (bit_offset + bit_length as usize - 1) / 8;
        let bit_shift = bit_offset % 8;

        // TODO check bounds!
        let bytes = &report[first_byte..=last_byte];

        let mut value = 0u32;
        for byte in 0..bytes.len() {
            // numbers are little-endian!
            value |= (bytes[byte as usize] as u32) << (8 * byte);
        }

        value >>= bit_shift;
        value &= !(0xFFFFFFFFu32 << bit_length);

        value
    }
}

#[derive(Debug)]
pub enum ReportType {
    Input(u32),
    // TODO ready for other types of report
    //
    // Output(u32),
    // Feature(u32),
}

impl ReportType {
    // TODO decoding of bit flags
}

#[cfg(test)]
mod test {
    use super::Report;

    #[test]
    fn extracts_single_bit_value() {
        let report: [u8; 1] = [0b1];
        let expected = 1;
        let actual = Report::extract_value(&report, 0, 1);

        assert_eq!(actual, expected);
        let report: [u8; 1] = [0b10];
        let expected = 1;
        let actual = Report::extract_value(&report, 1, 1);

        assert_eq!(actual, expected);

        assert_eq!(actual, expected);
        let report: [u8; 3] = [0b0, 0b0, 0b100];
        let expected = 1;
        let actual = Report::extract_value(&report, 18, 1);

        assert_eq!(actual, expected);

        let expected = 0;
        let actual = Report::extract_value(&report, 17, 1);

        assert_eq!(actual, expected);

        let expected = 0;
        let actual = Report::extract_value(&report, 19, 1);

        assert_eq!(actual, expected);
    }

    #[test]
    fn extracts_multi_bit_value() {
        let report: [u8; 1] = [0b101];
        let expected = 5;
        let actual = Report::extract_value(&report, 0, 3);

        assert_eq!(actual, expected);

        let report: [u8; 3] = [0b0, 0b0, 0b1010];
        let expected = 5;
        let actual = Report::extract_value(&report, 17, 3);

        assert_eq!(actual, expected);

        let report: [u8; 3] = [0b10000000, 0b10, 0b0];
        let expected = 5;
        let actual = Report::extract_value(&report, 7, 3);

        assert_eq!(actual, expected);

        let report: [u8; 3] = [0b10000000, 0b10, 0b00011];
        let expected = 0b11000000101;
        let actual = Report::extract_value(&report, 7, 11);

        assert_eq!(actual, expected);

        let report: [u8; 2] = [0b10, 0b1000_0000];
        let expected = 0b100_0000_0000_0001;
        let actual = Report::extract_value(&report, 1, 15);

        assert_eq!(actual, expected);
    }

    #[test]
    fn convert_any_bit_length_to_i32() {
        let actual = Report::signed((!27u8 + 1) as u32, 8);
        let expected = -27;

        println!("-27i32: {:032b}", -27i32);
        assert_eq!(actual, expected);

        let actual = Report::signed((!1u8 + 1) as u32, 8);
        let expected = -1;

        assert_eq!(actual, expected);

        let actual = Report::signed(1u8 as u32, 8);
        let expected = 1;

        assert_eq!(actual, expected);

        let actual = Report::signed(127u8 as u32, 8);
        let expected = 127;

        assert_eq!(actual, expected);

        let actual = Report::signed((!127u8 + 1) as u32, 8);
        let expected = -127;

        assert_eq!(actual, expected);
    }
}
