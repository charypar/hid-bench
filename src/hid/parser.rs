use std::collections::VecDeque;

use super::basic::{BasicItem, BasicItems, GlobalItem, InputItemData, LocalItem, MainItem};
use super::collection::{Collection, CollectionItem};
use super::input::Input;
use super::report::{Report, ReportType};

#[derive(Debug)]
pub struct Parser {
    collection: Collection<Report>,
}

impl Parser {
    pub fn new(basic_items: BasicItems<'_>) -> Self {
        Parser {
            collection: Self::read_items(basic_items),
        }
    }

    pub fn parse_input(&self, input: &[u8]) -> Collection<Vec<Input>> {
        self.collection.map(|report| report.parse(input))
    }

    // FIXME error handling
    fn read_items(basic_items: BasicItems) -> Collection<Report> {
        let global = GlobalItems::new();
        let local = LocalItems::new();
        let mut state_table = StateTable { global, local };

        let mut collection_stack: VecDeque<Collection<Report>> = VecDeque::new(); // current collection
        let mut bit_offset = 0u32;

        for item in basic_items {
            match item {
                BasicItem::Global(item) => {
                    Self::read_global_item(&mut state_table, item);
                }
                BasicItem::Local(item) => Self::read_local_item(&mut state_table, item),
                BasicItem::Main(item) => match item {
                    MainItem::Input(input) => Self::create_input_item(
                        &mut state_table,
                        &mut collection_stack,
                        &mut bit_offset,
                        input,
                    ),
                    // Output and feature items not yet implemented
                    MainItem::Output(_) => continue,
                    MainItem::Feature(_) => continue,
                    MainItem::Collection(c) => {
                        if state_table.local.usages.len() != 1 {
                            panic!("Too many usages for a collection");
                        }
                        let local_usage = state_table.local.usages[0];

                        // Start a new collection
                        let collection_type = c;
                        let usage =
                            Self::qualify_usage(&state_table.global.usage_page, &local_usage)
                                .expect("Bad usage item");

                        let collection = Collection {
                            collection_type,
                            usage,
                            designator_index: None,
                            string_index: None,
                            items: vec![],
                        };

                        // Make the collection the active one, main items will be pushed into it
                        collection_stack.push_back(collection);

                        // Clear the local state table
                        state_table.local = LocalItems::new();
                    }
                    MainItem::EndCollection => {
                        // close the collection and add it to its parrent collection items
                        if collection_stack.len() == 1 {
                            continue; // nothing to be done about the top level collection
                        }

                        let top = collection_stack.len() - 2;
                        let collection = collection_stack
                            .pop_back()
                            .expect("can't pop collection of a stack with items");

                        collection_stack[top]
                            .items
                            .push(CollectionItem::Collection(collection));
                    }
                    MainItem::Reserved => continue,
                },
                BasicItem::Reserved => continue,
            }
        }

        collection_stack.pop_front().expect("No collection found!")
    }

    // FIXME error handling
    fn read_global_item(state_table: &mut StateTable, item: GlobalItem) {
        match item {
            GlobalItem::UsagePage(up) => state_table.global.usage_page = Some(up),
            GlobalItem::LogicalMinimum(lm) => state_table.global.logical_minimum = Some(lm),
            GlobalItem::LogicalMaximum(lm) => state_table.global.logical_maximum = Some(lm),
            GlobalItem::PhysicalMinimum(pm) => state_table.global.physical_minimum = Some(pm),
            GlobalItem::PhysicalMaximum(pm) => state_table.global.physical_maximum = Some(pm),
            GlobalItem::UnitExponent(ue) => state_table.global.unit_exponent = Some(ue),
            GlobalItem::Unit(u) => state_table.global.unit = Some(u),
            GlobalItem::ReportSize(rs) => state_table.global.report_size = Some(rs),
            GlobalItem::ReportID(rid) => state_table.global.report_id = Some(rid),
            GlobalItem::ReportCount(rc) => state_table.global.report_count = Some(rc),
            GlobalItem::Push => {
                todo!("Item state table stack is not yet implemented")
            }
            GlobalItem::Pop => {
                todo!("Item state table stack is not yet implemented")
            }
            GlobalItem::Reserved => (),
        }
    }

    // FIXME error handling
    fn read_local_item(state_table: &mut StateTable, item: LocalItem) {
        match item {
            LocalItem::Usage(usage) => state_table.local.usages.push((None, Some(usage))),
            LocalItem::UsageMinimum(um) => state_table.local.usage_minimum = (None, Some(um)),
            LocalItem::UsageMaximum(um) => state_table.local.usage_maximum = (None, Some(um)),
            LocalItem::ExtendedUsage(up, usage) => {
                state_table.local.usages.push((Some(up), Some(usage)))
            }
            LocalItem::ExtendedUsageMinimum(up, um) => {
                state_table.local.usage_minimum = (Some(up), Some(um))
            }
            LocalItem::ExtendedUsageMaximum(up, um) => {
                state_table.local.usage_maximum = (Some(up), Some(um))
            }
            LocalItem::Delimiter(_) => todo!("Delimiters are not yet implemented"),
            // Strings and designators not yet implemented
            LocalItem::DesignatorIndex(di) => state_table.local.designator_index = Some(di),
            LocalItem::DesignatorMinimum(dm) => state_table.local.designator_minimum = Some(dm),
            LocalItem::DesignatorMaximum(dm) => state_table.local.designator_maximum = Some(dm),
            LocalItem::StringIndex(si) => state_table.local.string_index = Some(si),
            LocalItem::StringMinimum(sm) => state_table.local.string_minimum = Some(sm),
            LocalItem::StringMaximum(sm) => state_table.local.string_maximum = Some(sm),
            LocalItem::Reserved => (),
        }
    }

    // FIXME error handling!
    fn create_input_item(
        state_table: &mut StateTable,
        collection_stack: &mut VecDeque<Collection<Report>>,
        bit_offset: &mut u32,
        input: InputItemData,
    ) {
        let report_type = ReportType::Input(input.data);
        let usage_page = state_table.global.usage_page;

        let usages = state_table
            .local
            .usages
            .iter()
            .map(|usage| {
                Self::qualify_usage(&usage_page, usage).expect("Missing usage page for input item")
            })
            .collect();
        let usage_maximum = Self::qualify_usage(&usage_page, &state_table.local.usage_maximum);
        let usage_minimum = Self::qualify_usage(&usage_page, &state_table.local.usage_minimum);

        let report_size = state_table
            .global
            .report_size
            .expect("Missing report size for input item");
        let report_count = state_table
            .global
            .report_count
            .expect("Missing report size for input item");

        let logical_minimum = state_table
            .global
            .logical_minimum
            .expect("Missing logical minimum for input item");
        let logical_maximum = state_table
            .global
            .logical_maximum
            .expect("Missing logical minimum for input item");

        let physical_minimum = state_table
            .global
            .physical_minimum
            .unwrap_or(logical_minimum);
        let physical_maximum = state_table
            .global
            .physical_minimum
            .unwrap_or(logical_maximum);

        let report = Report {
            report_type,
            usages,
            usage_minimum,
            usage_maximum,
            bit_offset: *bit_offset as usize,
            report_id: state_table.global.report_id,
            report_size,
            report_count,
            logical_minimum,
            logical_maximum,
            physical_minimum,
            physical_maximum,
            unit: state_table.global.unit,
            unit_exponent: state_table.global.unit_exponent,
        };

        let top = collection_stack.len() - 1;
        collection_stack[top]
            .items
            .push(CollectionItem::Item(report));

        *bit_offset += report_count * report_size;
        state_table.local = LocalItems::new();
    }

    // FIXME error handling
    fn qualify_usage(
        usage_page: &Option<u16>,
        usage: &(Option<u16>, Option<u16>),
    ) -> Option<(u16, u16)> {
        match (usage_page, usage) {
            (_, (None, None)) => None,
            (_, (Some(up), Some(us))) => Some((*up, *us)),
            (Some(up), (None, Some(us))) => Some((*up, *us)),
            (None, (None, Some(_))) => panic!("Missing usage page"),
            _ => panic!("Missing usage"),
        }
    }
}

struct StateTable {
    global: GlobalItems,
    local: LocalItems,
}

struct GlobalItems {
    usage_page: Option<u16>,
    logical_minimum: Option<i32>,
    logical_maximum: Option<i32>,
    physical_minimum: Option<i32>,
    physical_maximum: Option<i32>,
    unit_exponent: Option<u32>,
    unit: Option<u32>,
    report_size: Option<u32>,
    report_id: Option<u8>,
    report_count: Option<u32>,
}

impl GlobalItems {
    fn new() -> Self {
        Self {
            usage_page: None,
            logical_minimum: None,
            logical_maximum: None,
            physical_minimum: None,
            physical_maximum: None,
            unit_exponent: None,
            unit: None,
            report_size: None,
            report_id: None,
            report_count: None,
        }
    }
}

struct LocalItems {
    usages: Vec<(Option<u16>, Option<u16>)>,   // page, usage
    usage_minimum: (Option<u16>, Option<u16>), // page, usage
    usage_maximum: (Option<u16>, Option<u16>), // page, usage
    designator_index: Option<u32>,
    designator_minimum: Option<u32>,
    designator_maximum: Option<u32>,
    string_index: Option<u32>,
    string_minimum: Option<u32>,
    string_maximum: Option<u32>,
    // TODO support delimiter
}

impl LocalItems {
    fn new() -> Self {
        Self {
            usages: vec![],
            usage_minimum: (None, None),
            usage_maximum: (None, None),
            designator_index: None,
            designator_minimum: None,
            designator_maximum: None,
            string_index: None,
            string_minimum: None,
            string_maximum: None,
        }
    }
}

#[cfg(test)]
mod test {
    use super::super::BasicItems;
    use super::Parser;

    const JOYSTICK: [u8; 101] = [
        0x5, 0x1, 0x9, 0x4, 0xa1, 0x1, 0x9, 0x1, 0xa1, 0x0, 0x5, 0x1, 0x9, 0x30, 0x9, 0x31, 0x15,
        0x0, 0x26, 0xff, 0x3, 0x75, 0xa, 0x95, 0x2, 0x81, 0x2, 0x9, 0x35, 0x15, 0x0, 0x26, 0xff,
        0x0, 0x75, 0x8, 0x95, 0x1, 0x81, 0x2, 0x9, 0x32, 0x9, 0x36, 0x15, 0x0, 0x26, 0xff, 0x0,
        0x75, 0x8, 0x95, 0x2, 0x81, 0x2, 0x5, 0x9, 0x19, 0x1, 0x29, 0xe, 0x15, 0x0, 0x25, 0x1,
        0x75, 0x1, 0x95, 0xe, 0x81, 0x2, 0x5, 0x1, 0x9, 0x39, 0x15, 0x1, 0x25, 0x8, 0x35, 0x0,
        0x46, 0x3b, 0x1, 0x66, 0x14, 0x0, 0x75, 0x4, 0x95, 0x1, 0x81, 0x42, 0x75, 0x2, 0x95, 0x1,
        0x81, 0x1, 0xc0, 0xc0,
    ];

    #[test]
    fn parses_basic_report_descriptor_items() {
        let basic_items = BasicItems::new(&JOYSTICK);
        println!("{:x?}", basic_items);
        println!("{:?}", basic_items.collect::<Vec<_>>());
    }

    #[test]
    fn parses_a_report_descriptor() {
        let basic_items = BasicItems::new(&JOYSTICK);
        let parser = Parser::new(basic_items);

        println!("{:#?}", parser);
    }

    #[test]
    fn parses_an_input_report() {
        let basic_items = BasicItems::new(&JOYSTICK);
        let parser = Parser::new(basic_items);

        let input_report = [0u8; 64];
        let input = parser.parse_input(&input_report);

        println!("{:#?}", input);
    }
}
