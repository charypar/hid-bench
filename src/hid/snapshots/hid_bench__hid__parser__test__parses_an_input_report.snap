---
source: src/hid/parser.rs
expression: input
---
Collection {
    collection_type: Application,
    usage: (
        1,
        4,
    ),
    designator_index: None,
    string_index: None,
    items: [
        Collection(
            Collection {
                collection_type: Physical,
                usage: (
                    1,
                    1,
                ),
                designator_index: None,
                string_index: None,
                items: [
                    Item(
                        [
                            Input {
                                usage: (
                                    1,
                                    48,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                            Input {
                                usage: (
                                    1,
                                    49,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                        ],
                    ),
                    Item(
                        [
                            Input {
                                usage: (
                                    1,
                                    53,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                        ],
                    ),
                    Item(
                        [
                            Input {
                                usage: (
                                    1,
                                    50,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                            Input {
                                usage: (
                                    1,
                                    54,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                        ],
                    ),
                    Item(
                        [
                            Input {
                                usage: (
                                    9,
                                    1,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    2,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    3,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    4,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    5,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    6,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    7,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    8,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    9,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    10,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    11,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    12,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    13,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                            Input {
                                usage: (
                                    9,
                                    14,
                                ),
                                value: Bool(
                                    false,
                                ),
                            },
                        ],
                    ),
                    Item(
                        [
                            Input {
                                usage: (
                                    1,
                                    57,
                                ),
                                value: UInt(
                                    0,
                                ),
                            },
                        ],
                    ),
                ],
            },
        ),
    ],
}
